[RFC][WIP] Make DEVICE init and device_get dt aware

galak opened the pull request:
This is prototype based on conversations at ELC-E 2019 on how we can make the device api be dt aware.  

The idea here is that we'd use a unique identifier (Zephyr Device ID/ZDID) for each device tree node that would correspond to how we name the `struct device` for that device.  We can then generate various DT defines like `DT_INST_0_SILABS_GECKO_I2C_ZDID`.  These defines allow us to construct direct access to the `struct device`.  This PR show changes to the silabs i2c gecko driver and a hacked up change to the i2c_scanner to show but the changes on the driver side (how a `DEVICE_AND_API_INIT_DT` would look like, and how `device_get_dt_binding` would work.

There's at least one issue that needs to be figured out on how / where to do an `extern struct device __device_<ZDID>`.

galak committed:
dts: gen_defines: generate ZDID for devices

Introduce a Zephyr Device ID value that we can use to reference DT nodes
by a number....

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

add zdid support

galak committed:
device: Introduce DT based init of device struct

Introduce DEVICE_AND_API_INIT_DT to uses DTS ZDID construct to define
the name of the struct in a consistent manor.  We also remove the static
define from the 'struct device' since we will be able to directly
provide access to it.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
i2c: gecko: convert to use new DEVICE_AND_API_INIT_DT

make device driver DT api friendly.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
samples: i2c_scanner: show example of device_get_dt_binding

Hack up device_get_dt_binding, its really device_get_dt_binding_by_inst.
Need to figure out how to handle the extern of __device_ZDID.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak commented:
Benefits of this approach:
* ability to remove the 'label' property from DT nodes [ been a long term goal]
* ability to remove the device name and thus string storage
* code and perf improvement for `device_get_dt_*` -- its just a pointer dereference now!

erwango commented:
> ability to remove the 'label' property from DT nodes [ been a long term goal]

I'm still not convinced this is something desirable.
Or at least I think information held in 'label' is a valid an pertinent information.
Referencing to instance, ie index of the enabled peripheral is actually random from board configuration point of view, and so referencing to instance from application perspective is misleading. Besides, it means nothing neither from SoC documents point of view, (ref mans and so on).
If label is removed, we need to be able to get access from the code to device node name, which hold same information.

pabigot commented on scripts/dts/gen_defines.py:
```diff
@@ -101,6 +107,7 @@ def main():
             write_existence_flags(node)
 
             out_dev(node, "ZDID", node_to_zdid[node])
+            print("extern struct device __device_{};".format(node_to_zdid[node]), file=dev_extern_file)
```
It would be nice if this include some identification:
```
...
extern struct device __device_4;   /* SILABS_GETCKO_I2C#0 */
```
or something else so we could catch the declaration name using grep.  I'm imagining cases where I have a pointer, look it up in a linker map, and want to know what it belongs to.

pabigot reviewed (commented):
Too bad I'm not there; this is exactly the sort of thing I'm interested in.

In principle I think this is great.  I'd intended to use the ordinal generated by #19454 for this, with the thought of providing a lookop table rather than external name, but this works.  A benefit of using the ordinal is that we could put all the device instance records into a linker section in dependency order, which could simplify initialization.

One concern is that the generated names like `DT_INST_0_NORDIC_NRF_SAADC_ZDID` intrude on the binding namespace, since it conflicts with a binding property named `ZDID`.  We've made progress to eliminating that by using prefixes instead; it's sad to introduce a new case.

To @erwango's point I think keeping the `label` property is fine: it should just be used for its defined purpose "human readable string describing a device" rather than as a handle that identifies the device within a registry.  Though I don't know how we deal with all the `device_get_binding(DT_INST_0_FOO_LABEL)` uses that already exist.

ulfalizer commented on scripts/dts/gen_defines.py:
```diff
@@ -39,6 +41,11 @@ def main():
+
+    print("#ifndef GENERATED_DTS_DEV_EXTERN_H", file=dev_extern_file)
+    print("#define GENERATED_DTS_DEV_EXTERN_H", file=dev_extern_file)
+    print("#ifndef _ASMLANGUAGE", file=dev_extern_file)
```
Can despam a bit with

```python
print("#ifndef GENERATED_DTS_DEV_EXTERN_H\n"
      "#define GENERATED_DTS_DEV_EXTERN_H\n"
      "#ifndef _ASMLANGUAGE", file=dev_extern_file)
```

or

```python
print("""\
#ifndef GENERATED_DTS_DEV_EXTERN_H
#define GENERATED_DTS_DEV_EXTERN_H
#ifndef _ASMLANGUAGE\
""", file=dev_extern_file)
```

or the like.

ulfalizer commented on scripts/dts/gen_defines.py:
```diff
@@ -53,6 +60,23 @@ def main():
     active_compats = set()
 
+    zdid = 1
+    node_to_zdid = {}
```
`node2zdid`, to be consistent with other lookup tables.

ulfalizer commented on scripts/dts/gen_defines.py:
```diff
@@ -53,6 +60,23 @@ def main():
+
+            node_to_zdid[node] = zdid
+            zdid = zdid + 1
+
```
Could move into a separate function (`_init_zdid()` or the like). Would be nice if we could keep `main()` about the same size forever.

ulfalizer commented on scripts/dts/gen_defines.py:
```diff
@@ -120,6 +149,8 @@ def parse_args():
     parser.add_argument("--header-out", required=True,
                         help="path to write header to")
+    parser.add_argument("--device-extern-out", required=True,
+                        help="path to write header to")
```
Needs a better help string.

galak replied:
Removed it.

ulfalizer commented on scripts/dts/gen_defines.py:
```diff
@@ -107,6 +134,8 @@ def main():
         out("FLASH_AREA_NUM", flash_index)
 
+    print("#endif /* _ASMLANGUAGE */", file=dev_extern_file)
+    print("#endif /* GENERATED_DTS_DEV_EXTERN_H */", file=dev_extern_file)
```
Could despam.

galak replied:
Removed it.

ulfalizer reviewed (commented):
Some nits. Need to look into how this stuff works atm in more detail.

ulfalizer commented on scripts/dts/gen_defines.py:
```diff
@@ -82,6 +106,9 @@ def main():
             write_bus(node)
             write_existence_flags(node)
 
+            out_dev(node, "ZDID", node_to_zdid[node])
```
Do we need to generate `..._ZDID` macros for all the device aliases and stuff too, like `out_dev()` does? If not, maybe the ID could be special-cased and just generate the needed macro.

The format of the macro could be modified then too, so that it doesn't clash with regular properties.

What do you think about having

```
DT_DEV_ID_INST_##inst##_##dev_name##
```

instead of

```
DT_INST_##inst##_##dev_name##_ZDID
```

Something more plain-English and guessable like "DEV_ID" seems nicer than "ZDID". Already know we're dealing with Zephyr. ;)

galak commented:
> I'm still not convinced this is something desirable.
> Or at least I think information held in 'label' is a valid an pertinent information.
> Referencing to instance, ie index of the enabled peripheral is actually random from board configuration point of view, and so referencing to instance from application perspective is misleading. Besides, it means nothing neither from SoC documents point of view, (ref mans and so on).
> If label is removed, we need to be able to get access from the code to device node name, which hold same information.

The application can reference via instance, alias, path.  I just did it via instance in here.

so we can have:

`device_get_dt_by_path(SOC_I2C_4000C400)`
`device_get_dt_by_alias(I2C_0)`
etc.

ulfalizer commented:
> The application can reference via instance, alias, path. I just did it via instance in here.
> 
> so we can have:
> 
> `device_get_dt_by_path(SOC_I2C_4000C400)` [maps to /soc/i2c@4000C400]
> `device_get_dt_by_alias(I2C_0)`
> etc.

Seems pretty neat.

Guess the aliases generated by `out_dev()` are needed in that case. Would expand to a different `*_ZDID` alias depending on how you look it up, and then that gets resolved to the right `device` struct.

Still wonder if there's a nicer name than `ZDID`. Could make it `*__ZDID` (or just `*__ID`) maybe, corresponding to `_id = ...`, though it might be overkill.

erwango commented:
> The application can reference via instance, alias, path. I just did it via instance in here:
```
device_get_dt_by_path(SOC_I2C_4000C400) [maps to /soc/i2c@4000C400]
```

I don't think this scales, peripheral address is not used in ref manuals (it is not even available) nor available on most of board documentation. User have to go and check:
-  CMSIS file (in which it is not even directly available `#define GPIOH_BASE            (AHB1PERIPH_BASE + 0x1C00UL)`).
- .dtsi (which are split in multiple files, so need to parse several files) 


```
device_get_dt_by_alias(I2C_0)
```

So we need to define an alias for each peripheral instance... Not sure this scales either.

Anyway we can use node name (ie phandle)?

mbolivar commented:
> Anyway we can use node name (ie phandle)?

The vision for this that I discussed with Kumar was that "it's just a pointer dereference!" makes phandles "just work": from `prop = <&gpioM ...>` you can translate the `&gpioM` phandle to `&__device_N` in the generated header "right away", since the script knows the ID number for the node pointed to by the phandle.

pabigot commented:
> Anyway we can use node name (ie phandle)?

The node name isn't the phandle, which isn't text; the generic syntax is:
```
    [label:] node-name[@unit-address] { 
      [phandle = <*int32*>;]
      ... 
   }
```
The label is what's used as the reference in things like `<&gpiof 1 0>`, and it evaluates (in blob representations) to the integer that is the (usually implicit) `phandle` property.  The `node-name` need only be unique within its parent, so it's not a good choice.

Nomenclature nit-picking aside, I agree that `device_get_dt_by_label()` would be useful, if we can figure out how to name it so it's not confused with the `label` property.

pabigot commented:
> The vision for this that I discussed with Kumar was that "it's just a pointer dereference!" makes phandles "just work": from `prop = <&gpioM ...>` you can translate the `&gpioM` phandle to `&__device_N` in the generated header "right away", since the script knows the ID number for the node pointed to by the phandle.

This makes sense.

Something to consider is that AFAIK to this point the only way to get a `struct device *` was to use `device_get_binding(LABEL)`, and the result is null if the device exists but does not function.

`device_get_dt_binding*()` appears to return a pointer to a `struct device` whether or not that device works (and will fail to build if the referenced device is not linked in).  This is a "good thing" as it enables progress on #19448 where we need to be able to turn off devices, or turn on devices that weren't enabled at boot.  However, we do need to take care that driver access is made safe against invocations where the pointer was obtained without validating the device or before the device was started.  That may require a tree-wide inspection, or something injected at the driver wrapper level.

ulfalizer commented on scripts/dts/gen_defines.py:
```diff
@@ -53,6 +60,23 @@ def main():
+                continue
+
+            node_to_zdid[node] = zdid
+            zdid = zdid + 1
```
Nit: Can do `zdid += 1`.

galak commented:
> Too bad I'm not there; this is exactly the sort of thing I'm interested in.
> 
> In principle I think this is great. I'd intended to use the ordinal generated by #19454 for this, with the thought of providing a lookop table rather than external name, but this works. A benefit of using the ordinal is that we could put all the device instance records into a linker section in dependency order, which could simplify initialization.

Agreed, the intent was the assign the numbers in that order, I was just being quick with giving some numbering system to show something work.  But totally agree this should be 

> One concern is that the generated names like `DT_INST_0_NORDIC_NRF_SAADC_ZDID` intrude on the binding namespace, since it conflicts with a binding property named `ZDID`. We've made progress to eliminating that by using prefixes instead; it's sad to introduce a new case.

Fair, I thought about that and agree.  This will get fixed.
 
> To @erwango's point I think keeping the `label` property is fine: it should just be used for its defined purpose "human readable string describing a device" rather than as a handle that identifies the device within a registry. Though I don't know how we deal with all the `device_get_binding(DT_INST_0_FOO_LABEL)` uses that already exist.

The hope is to remove the string at some point since it would reduce footprint.

pabigot commented:
> > To @erwango's point I think keeping the `label` property is fine: it should just be used for its defined purpose "human readable string describing a device" rather than as a handle that identifies the device within a registry. Though I don't know how we deal with all the `device_get_binding(DT_INST_0_FOO_LABEL)` uses that already exist.
> 
> The hope is to remove the string at some point since it would reduce footprint.

I was thinking of a Kconfig option that would cause the value of the `label` property to be stored in the device `device_config` structure.  This would allow access to that property from user code, and would contribute to solving the problem of displaying device information in a human-readable way at runtime.

tbursztyka commented on include/device.h:
```diff
@@ -9,6 +9,7 @@
 #define ZEPHYR_INCLUDE_DEVICE_H_
 
 #include <kernel.h>
+#include <dts_device_extern.h>
```
Why not instead including it as the other generated_dts_*.h files?

galak replied:
Not sure I follow.

tbursztyka replied:
including it in generated_dts_board.h
I find it awkward to have this new generated header in a public header file such as device.h

galak replied:
There's an issue in that generated_dts_board.h is just defines, putting `externs` in there causes issues with the linker scripts as they include and I think utilize defines from generated_dts_board.h

mbolivar replied:
can't we `#ifndef _LINKER_SCRIPT` our way around that?

pabigot replied:
Or advance to putting them all into a linker section in ZDID order, and indexing into the resulting array with ZDID.  A benefit of that is that the opaque handle could be a `u16_t`, we could retain `device_get_binding(ID)` with a change in argument signature (or a parallel function with different name), and we don't have to worry as much about people trying to use uninitialized devices just because they got the pointer to one somehow.  Then there's just:
```
extern struct device *__devices[];
#define device_get_dt_binding(BLAH) __devices[MAGIC_GET_ZDID(BLAH)]
```
somewhere.

(I assume one ordinal reserved for "no-such-device", and no Zephyr image would have more than 65534 devices.  254 seems too small.)

galak replied:
> can't we `#ifndef _LINKER_SCRIPT` our way around that?

possibly, didn't know this ifdef existed.

mbolivar replied:
> possibly, didn't know this ifdef existed.

The name is just a suggestion, I'm saying we can make sure a macro is defined only when running cpp over the ld scripts and use that here

pabigot commented:
> > ... use the ordinal generated by #19454 ...
> Agreed, the intent was the assign the numbers in that order

That would allow us to use the ordinal (`ZDID`) as the opaque device handle, with `-1` to represent an unknown device (or 0 if we muck with the device array indexing, so references initialize to "no known device").

Then there's still a barrier of going through `device_get_binding_by_id()` to get a validated  `struct device *`.

tbursztyka reviewed (commented):
I am not sure we want to separate the base identifier of a device and its instance number.
First because it will require to deprecate device_get_binding(), second because its replacement will require 2 parameters (ok that's a minor issue, but it adds up a bit of complexity: before 1 param, now 2... to do the same thing).

galak commented on samples/drivers/i2c_scanner/src/main.c:
```diff
@@ -9,13 +9,19 @@
+#define device_get_dt_binding_zdid(zdid) _CONCAT(&__device_, zdid)
+
+#define device_get_dt_binding(name, inst) \
+device_get_dt_binding_zdid(DT_INST_##inst##_##name##_ZDID) \
```
avoid it how?

galak replied:
first, what would be better?

tbursztyka replied:
> avoid it how?

Exactly my question :) 

I understand the idea behind this patches, however it's changing how user will get the bindings. It could be done directly with a static string being the same as the dts node label ("SPI_0" for instance). That was simple.

Also doing it that way was closely down to the board the user is using. He probably does not know which controller runs on the SoC, he has a board that exposes "spi_0, spi_1, uart_0, uart_1, uart_2" and it is then direct: device_get_binding("UART_0"). One can relate it directly to the board. 

Now it would require the user to know the controller, and the instance he needs, exposing then the instance concept to the user at the same time then... not sure it's wise.

Looks like I am in line with @erwango here.

mbolivar replied:
Keeping the labels available under a default y kconfig would let users keep doing this for upstream samples but still let us get rid of the labels to save ROM when needed, agreed @tbursztyka ?

tbursztyka replied:
When it comes to name, I think we want to get rid of them entirely. They are really useless.
(only maybe in debug mode, inside the actual driver, but nowhere else).

Instead we could generate BOARD_UART_0, BOARD_SPI_1, ... macros which in fact would hold the dev name (which is being build with the ZDID etc.. on this rfc). device_get_binding() would then call DEVICE_GET() with this, and since dts_device_extern.h already exposes the extern device pointers: done.

@galak That could be your solution to avoid fully deprecating device_get_binding(), though it would change it's signature (it would become a macro calling DEVICE_GET() macro).

galak replied:
I expect several `device_get` APIs that would be new.  We can continue to support `device_get_binding()` via string.  I'd expect we'd have:

`device_get_by_dt_label(LABEL) [this is the node label, not the 'label' property]`
`device_get_by_dt_alias(ALIAS)`
`device_get_by_dt_path(PATH)`
`device_get_by_dt_inst(NAME, ID)` [not sure if this one is useful]

tbursztyka replied:
ok sure, device_get_binding() would use device_get_by_dt_label then.

jfischer-phytec-iot replied:
>device_get_by_dt_inst(NAME, ID) [not sure if this one is useful]

If the NAME is type of the device and ID the index, then it would allow the same feature as suggested in #21709

galak commented on scripts/dts/gen_defines.py:
```diff
@@ -101,6 +101,9 @@ def main():
 
             out_dev(node, "ZDID", node_to_zdid[node])
+            print("#ifndef _ASMLANGUAGE", file=header_file)
+            print("extern struct device __device_{};".format(node_to_zdid[node]), file=header_file)
```
Will need @ulfalizer help with that, but should be doable.

galak committed:
device: change struct device to non-static

As a pre-cursor to use DT to get access to the struct device directly we
need to remove the static init of a struct device.h

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak commented on scripts/dts/gen_defines.py:
```diff
@@ -553,6 +576,9 @@ def write_phandle_val_list_entry(node, entry, i, ident):
         out_dev_s(node, ctrl_ident, entry.controller.label)
 
+        out_dev(node, ctrl_ident + "_ZDID", node_to_zdid[entry.controller])
+
```
removed.

galak committed:
linker: move where we define _LINKER and _ASMLANGUAGE

Move _LINKER and _ASMLANGUAGE to target.cmake because of how we pick the
linker script that might be used.  This way regardless of how or where a
linker.ld gets included we will always set _LINKER & _ASMLANGUAGE (so
any header that needs check based on those defines they can,
specifically generated_dts_board.h)

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

b0661 commented on scripts/dts/gen_defines.py:
```diff
@@ -53,6 +54,28 @@ def main():
+    zdid = 1
+    node2zdid = {}
+
+    # Assign a ZDID (Zephyr Device ID) to all nodes before we start
```
Device id should be assigned by edts (not gen_defines.py). 

Device id seems to become a vital part of device handling with DTS in Zephyr. Having device id in edts makes it instantaneously available to any tool (e.g. code generation) that works on DTS in Zephyr. Otherwise the device id has to be extracted from the generated header file, which is just a waste of (maintenance) effort.

tbursztyka commented:
@pabigot 
> Something to consider is that AFAIK to this point the only way to get a struct device * was to use device_get_binding(LABEL), and the result is null if the device exists but does not function.

That was a quick hack to solve the issue when device init fails.
But we need to fix this (in another PR) in a better way (either storing the actual init error for a device, or having just a bit array: when 0, it's non functioning. Stored in kernel/device.c, not struct device. driver APIs or device driver would obligatory check this etc... just wet finger thoughts), since writing in struct device voids the possibility to get it always in ROM. (and if struct device was never touched at all, it would be possible to make it const also).
